### PR TITLE
LatentMoE implementation

### PR DESCRIPTION
**Pull request by AkshitaB**
## Summary

  This PR adds LatentMoE support to the original MoE and MoE v2 implementations.

  Instead of running routed experts on the model’s hidden representation, LatentMoE wraps the routed expert path with two shared linear projections:

  1. Project routed token payloads into a smaller latent representation: d → ℓ
  2. Perform expert dispatch, expert computation, and combine in latent space
  3. Project the combined routed output back into model space: ℓ → d

  The router continues to compute gating decisions from the original model-space representation. Only the routed payload and routed expert computation use
  the latent dimension.

  Shared experts, transformer norms, and residual operations remain in model space.

  ## Configuration

  ### Original MoE

  For the original MoE implementation, add latent_moe to MoEConfig:
```
  from olmo_core.nn.moe import LatentMoEConfig, MoEConfig

  moe = MoEConfig(
      ...,
      latent_moe=LatentMoEConfig(
          latent_dim=1024,
      ),
  )
```
  The router remains model-width, while the routed experts are built automatically with latent_dim.

  ### MoE v2

  For MoE v2, configure the router with the model dimension and the routed experts with the latent dimension:

```
  from olmo_core.nn.moe import LatentMoEConfig
  from olmo_core.nn.moe.v2.routed_experts import RoutedExpertsConfig
  from olmo_core.nn.moe.v2.router import MoERouterConfigV2

  d_model = 4096
  latent_dim = 1024

  block = OLMoDDPTransformerBlockConfig(
      ...,
      latent_moe=LatentMoEConfig(
          latent_dim=latent_dim,
      ),
      routed_experts_router=MoERouterConfigV2(
          d_model=d_model,
          ...,
      ),
      routed_experts=RoutedExpertsConfig(
          d_model=latent_dim,
          ...,
      ),
  )
```

  The following dimensional constraints are enforced:

  - 0 < latent_dim < d_model
  - routed_experts_router.d_model == d_model
  - routed_experts.d_model == latent_dim
  - Shared expert and shared router dimensions remain equal to d_model

  Leave latent_moe=None to use the standard MoE architecture.

  ## Optional pre-up-projection normalization

  Normalization before the ℓ → d projection is disabled by default.

  Enable it with:
```
  latent_moe=LatentMoEConfig(
      latent_dim=1024,
      up_proj_input_norm_enabled=True,
  )
```

  When enabled without an explicit configuration, a bias-free RMSNorm is used.

  A different normalization can be configured explicitly:

```
  from olmo_core.nn.layer_norm import LayerNormConfig, LayerNormType

  latent_moe=LatentMoEConfig(
      latent_dim=1024,
      up_proj_input_norm_enabled=True,
      up_proj_input_norm=LayerNormConfig(
          name=LayerNormType.rms,
          eps=1e-6,
          bias=False,
      ),
  )
```

  Projection biases are disabled by default and can be enabled with bias=True.

  ## Supported execution paths

  LatentMoE is integrated with:

  - Original MoE and dropless MoE
  - MoE v2 without expert parallelism
  - Synchronous and asynchronous expert parallelism
  - Rowwise, rowwise-wave, and DeepEP v2 backends
  - Hybrid MoE combined-forward paths
  - Tensor, expert, context, data, and pipeline parallel execution

  Routing remains model-width in every path, while routed communication and expert computation use latent_dim.

### CODE REVIEW

File: src/olmo_core/nn/moe/moe.py [hunk 1]
```diff
@@ -231,25 +311,42 @@ def forward(
         :returns: The output of the MoE layer, the optional load-balancing loss, and the optional
             router Z-loss.
         """
+        routed_x = self.prepare_routed_input(x)
         expert_weights, expert_indices, batch_size_per_expert, router_aux_loss = self.router(
-            x, loss_div_factor=loss_div_factor
+            routed_x, loss_div_factor=loss_div_factor
```
Comment (jacob-morrison) [inline src/olmo_core/nn/moe/moe.py]: IIRC in the latentMoE paper and in Kimi K3 the router sees the full tokens, and then uses that info to route the down-projected tokens

File: src/olmo_core/nn/ddp/block.py [hunk 2]
```diff
@@ -1356,6 +1448,18 @@ def _prepare_moe_input(self, x: torch.Tensor) -> torch.Tensor:
...
+        if self.latent_down_proj is None:
+            return model_input
+        return self.latent_down_proj(model_input)
+
+    def _restore_routed_moe_output(self, routed_output: torch.Tensor) -> torch.Tensor:
+        if self.latent_up_proj is None:
+            return routed_output
+        if self.latent_up_proj_input_norm is not None:
+            routed_output = self.latent_up_proj_input_norm(routed_output)
+        return self.latent_up_proj(routed_output)
```
[the remark this answers is not in the document]
Comment (jacob-morrison) [reply src/olmo_core/nn/ddp/block.py:1463]: my codex independently agreed with this, so this seems reasonable to me!

**Review by jacob-morrison — APPROVED** [review]
LGTM! One comment just for type safety, but otherwise seems good
